README.md
# fluent-plugin-bunyan

[Fluentd](https://fluentd.org/) parser plugin to process logs from applications that write [bunyan](https://github.com/gildas/go-logger/blob/master/trentm/node-bunyan) logs.

The plugin requires [Fluentd](https://fluentd.org/) 1.0 at least.

|  |   |
---|---|
master | [![Build Status](https://dev.azure.com/keltiek/gildas/_apis/build/status/gildas.fluent-plugin-bunyan?branchName=master)](https://dev.azure.com/keltiek/gildas/_build/latest?definitionId=4&branchName=master)
dev | [![Build Status](https://dev.azure.com/keltiek/gildas/_apis/build/status/gildas.fluent-plugin-bunyan?branchName=dev)](https://dev.azure.com/keltiek/gildas/_build/latest?definitionId=4&branchName=dev)

## Installation

Pretty straight forward:  
```console
$ gem install fluent-plugin-bunyan
```

## Configuration

Simply add the parser to the [Fluentd](https://fluentd.org/) configuration file, for example:

```xml
<source>
  @type tail # say the logs come from stdout
  <parse>
    @type bunyan
  </parse>
</source>
```

## Testing

To run the unit tests, make sure you have the [Fluentd](https://fluentd.org/) gem installed and run:

```console
rake test
```

## Copyright

* Copyright(c) 2020- Gildas Cherruel
* License
  * MIT


CHANGELOG.md
# Changelog

This file contains changes throughout the life of this project.

### Development
[Full Changelog](https://github.com/gildas/fluent-plugin-bunyan/compare/v0.0.4...dev)

New Features:
* None Yet

Bug Fixes:  
* None Yet

### 0.0.4 / 2020-03-17
[Full Changelog](https://github.com/gildas/fluent-plugin-bunyan/compare/v0.0.3...v0.0.4)

Bug Fixes:
* JSON logs were not extracted properly when coming from Kubernetes

### 0.0.3 / 2020-03-15
[Full Changelog](https://github.com/gildas/fluent-plugin-bunyan/compare/v0.0.2...v0.0.3)

New Features:
* Not exactly a feature, but tests go through Azure Pipeline

Bug Fixes:
* Removed the requirement of rake version as per gem build warning
* The Fluent parser was not "required" properly besides tests passing
* License was still Apache on the README file

### 0.0.2 / 2020-03-10
[Full Changelog](https://github.com/gildas/fluent-plugin-bunyan/compare/v0.0.1...v0.0.2)

Bug Fixes:
* The GEM spec requires rake version 12.3.3 at least. 
  There is a vulnerability before rake 12.3.3: [CVE-2020-8130](https://github.com/advisories/GHSA-jppv-gw3r-w3q8).  
  We just should be on the safe side...

### 0.0.1 / 2020-03-10

Initial Release


test/plugin/test_parser_bunyan.rb
require "helper"
require "fluent/plugin/parser_bunyan.rb"

class BunyanParserTest < Test::Unit::TestCase
  setup do
    Fluent::Test.setup
  end

  CONFIG = %[
    pattern apache
  ]

  sub_test_case "parse" do
    test "can parse simple entry" do
      driver = create_driver(CONFIG)
      text = '{"hostname":"helloworld-go-helloworld-5649dfd7bd-k8954","level":30,"msg":"The message","name":"helloworld","pid":1,"scope":"main","tid":1,"time":"2020-03-10T07:08:41Z","topic":"main","v":0}'
      expected_time   = Time.utc(2020, 3, 10, 7, 8, 41).to_i
      expected_record = {
        "host"     => "helloworld-go-helloworld-5649dfd7bd-k8954",
        "name"     => "helloworld",
        "severity" => "info",
        "level"    => 30,
        "message"  => "The message",
        "topic"    => "main",
        "scope"    => "main",
        "pid"      => 1,
        "tid"      => 1,
      }
      driver.instance.parse(text) do |time, record|
        assert_equal(expected_record, record)
        assert_equal(expected_time,   time)
      end
    end

    test "can parse entry with custom properties" do
      driver = create_driver(CONFIG)
      text = '{"hostname":"helloworld-go-helloworld-5649dfd7bd-k8954","level":50,"msg":"Failed to be happy","name":"helloworld","pid":1,"reqid":"0ba850ff-1e1f-4bcf-8288-135927bb850f","scope":"start","tid":21,"time":"2020-03-10T15:04:19Z","topic":"server","v":0}'
      expected_time   = Time.utc(2020, 3, 10, 15, 4, 19).to_i
      expected_record = {
        "host"     => "helloworld-go-helloworld-5649dfd7bd-k8954",
        "name"     => "helloworld",
        "severity" => "error",
        "level"    => 50,
        "message"  => "Failed to be happy",
        "topic"    => "server",
        "scope"    => "start",
        "pid"      => 1,
        "tid"      => 21,
        "reqid"    => "0ba850ff-1e1f-4bcf-8288-135927bb850f",
      }
      driver.instance.parse(text) do |time, record|
        assert_equal(expected_record, record)
        assert_equal(expected_time,   time)
      end
    end

    test "can parse entry with complex custom properties" do
      driver = create_driver(CONFIG)
      text = '{"hostname":"helloworld-go-helloworld-5649dfd7bd-k8954","level":50,"msg":"Failed to be happy","name":"helloworld","pid":1,"err":{"text": "aouch", "code": 404},"reqid":"0ba850ff-1e1f-4bcf-8288-135927bb850f","scope":"start","tid":21,"time":"2020-03-10T15:04:19Z","topic":"server","v":0}'
      expected_time   = Time.utc(2020, 3, 10, 15, 4, 19).to_i
      expected_record = {
        "host"     => "helloworld-go-helloworld-5649dfd7bd-k8954",
        "name"     => "helloworld",
        "severity" => "error",
        "level"    => 50,
        "message"  => "Failed to be happy",
        "topic"    => "server",
        "scope"    => "start",
        "pid"      => 1,
        "tid"      => 21,
        "reqid"    => "0ba850ff-1e1f-4bcf-8288-135927bb850f",
        "err"      => {
          "code" => 404,
          "text" => "aouch",
        },
      }
      driver.instance.parse(text) do |time, record|
        assert_equal(expected_record, record)
        assert_equal(expected_time,   time)
      end
    end
  end

    test "can parse entry from Kubernetes" do
      driver = create_driver(CONFIG)
      text = '2020-03-10T07:08:41.710251038+09:00 stdout F {"hostname":"helloworld-go-helloworld-5649dfd7bd-k8954","level":30,"msg":"The message","name":"helloworld","pid":1,"scope":"main","tid":1,"time":"2020-03-10T07:08:41Z","topic":"main","v":0}'
      expected_time   = Time.utc(2020, 3, 10, 7, 8, 41).to_i
      expected_record = {
        "host"     => "helloworld-go-helloworld-5649dfd7bd-k8954",
        "name"     => "helloworld",
        "severity" => "info",
        "level"    => 30,
        "message"  => "The message",
        "topic"    => "main",
        "scope"    => "main",
        "pid"      => 1,
        "tid"      => 1,
      }
      driver.instance.parse(text) do |time, record|
        assert_equal(expected_record, record)
        assert_equal(expected_time,   time)
      end
    end

  private

  def create_driver(conf = CONFIG)
    Fluent::Test::Driver::Parser.new(Fluent::Plugin::BunyanParser).configure(conf)
  end
end


lib/fluent/plugin/parser_bunyan.rb
#
# Copyright 2020- Gildas Cherruel
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require "fluent/parser"

module Fluent
  module Plugin
    class BunyanParser < JSONParser
      Fluent::Plugin.register_parser("bunyan", self)

      def configure(conf)
        conf["time_format"] = "%iso8601"
        super(conf)
      end

      def parse(text)
        # Example of a log line we want to get stuff from:
        # 2020-03-16T22:14:05.710251038+09:00 stdout F {"UserAgent":"kube-probe/1.17","hostname":"helloworld-go-helloworld-5649dfd7bd-k8954","level":30,"msg":"request start: GET /","name":"helloworld","pid":1,"remote":"10.1.10.1:60114","reqid":"e4745ed6-82d4-4013-aeda-1d02bef42331","scope":"/","tid":17,"time":"2020-03-16T13:14:05Z","topic":"route","v":0}
        if text =~ /[^{]*({.*)/
          super(text.sub(/[^{]*({.*)/, '\1')) do |time, record|
            if record then
              record["severity"] = severity(record["level"] || 0)
              record["host"]     = record.delete("hostname")
              record["message"]  = record.delete("msg")

              record.delete("v")
            end
            yield time, record
          end
        end
      end

      def severity(level)
        return "fatal" if level >= 60
        return "error" if level >= 50
        return "warn"  if level >= 40
        return "info"  if level >= 30
        return "debug" if level >= 20
        return "trace"
      end

      def version
        return "0.0.4"
      end
    end
  end
end
